xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,299 @@
+    } catch (IOException e) {
+      logger.log(Level.SEVERE, "Error from readBootstrap", e);
+      return;
+    }
+    final List<Bootstrapper.ServerInfo> serverList = bootstrapInfo.getServers();
+    final Node node = bootstrapInfo.getNode();
+    if (serverList.isEmpty()) {
+      logger.log(Level.SEVERE, "No traffic director provided by bootstrap");
+      return;
+    }
+    final EventLoopGroup timeService = SharedResourceHolder.get(eventLoopGroupResource);
+
+    XdsClient.XdsClientFactory xdsClientFactory =
+        new XdsClient.XdsClientFactory() {
+          @Override
+          XdsClient createXdsClient() {
+            return new XdsClientImpl(
+                "",
+                serverList,
+                XdsClient.XdsChannelFactory.getInstance(),
+                node,
+                syncContext,
+                timeService,
+                new ExponentialBackoffPolicy.Provider(),
+                GrpcUtil.STOPWATCH_SUPPLIER);
+          }
+        };
+    xdsClientPool = new XdsClient.RefCountedXdsClientObjectPool(xdsClientFactory);
+    xdsClient = xdsClientPool.getObject();
+    setClientAndWatcher(port);
+  }
+
+  @VisibleForTesting
+  XdsClientWrapperForServerSds(int port, XdsClient xdsClient) {
+    this.logId = InternalLogId.allocate("GrpcServer", Integer.toString(port));
+    this.xdsClient = xdsClient;
+    setClientAndWatcher(port);
+  }
+
+  private void setClientAndWatcher(int port) {
+    xdsClient.watchListenerData(
+        port,
+        new XdsClient.ListenerWatcher() {
+          @Override
+          public void onListenerChanged(XdsClient.ListenerUpdate update) {
+            logger.log(
+                Level.INFO,
+                "Setting myListener from ConfigUpdate listener :{0}",
+                update.getListener().toString());
+            curListener = update.getListener();
+          }
+
+          @Override
+          public void onError(Status error) {
+            // In order to distinguish between IO error and resource not found, set curListener
+            // to null in case of NOT_FOUND
+            if (error.getCode().equals(Status.Code.NOT_FOUND)) {
+              curListener = null;
+            }
+            // TODO(sanjaypujare): Implement logic for other cases based on final design.
+            logger.log(Level.SEVERE, "ListenerWatcher in XdsClientWrapperForServerSds:{0}", error);
+          }
+        });
+  }
+
+  /**
+   * Locates the best matching FilterChain to the channel from the current listener and returns the
+   * DownstreamTlsContext from that FilterChain.
+   */
+  public DownstreamTlsContext getDownstreamTlsContext(Channel channel) {
+    if (curListener != null && channel != null) {
+      List<FilterChain> filterChains = curListener.getFilterChains();
+      int highestScore = -1;
+      FilterChain bestMatch = null;
+      for (FilterChain filterChain : filterChains) {
+        int curScore = getFilterChainMatchScore(filterChain.getFilterChainMatch(), channel);
```

Comment:
Can you give me some idea why this approach is more error prone? Errors are possible with computation of the score (or heuristics involved in assigning a score) but I don't see why `Collections.max` is better (except in cases where you have multiple max candidates and there is some logic about which one to pick). Also we don't need elaborate sort/select algorithm as I don't see the control plane sending more than one or two with complex selection criterion.

Reply:
it is maybe not as error prone as it is (although it is harder to read the code). when you try to modify, it can be very hard to keep everything aligned (how many methods here need to be modified to add new logic?). thus, those score has no meaning and it only makes reader/maintainer confused. if you implement it as comparator, it is easy to add new logic. this it doesn't have any performance penalty in terms of big O.
i tried to do the same thing before. and, i got same comment which I now agree.

Reply:
Okay, just to clarify: you are saying to ditch the whole idea of scores and just compare 2 filterChains for a given channel? So initialize the `Comparator` with the channel and then given 2 filterChains pick one? That's interesting and let me think about it. In any case I think is probably overkill for the number of filterChains we will be expected to compare (in most cases just 1?) But I'll definitely consider it

Reply:
`Collections.max` with comparator doesn't work here. There may not be a matching filter at all in which case we need to return null and that's not handled by `max`.

Reply:
post check is what i was expecting. this is relatively minor issue compare to the score based approach.

Reply:
@creamsoup Honestly I think the Comparator here is an overkill assuming the control plane will mostly give us a single filterChain matching our requirements. The multiple filterChains happens with Envoy (since it's a proxy) but most likely won't happen with proxyless gRPC. I still need to confirm with the control plane team and if that is confirmed I can even remove the current logic and error in case of multiple filters and verify the single filter returned matches what I need.

Reply:
i know it can be tedious to covert but it is not hard nor overkill (compare to score based) while it makes the code more maintainable/readable.

Reply:
I spent a whole afternoon on it and I realized I still have to compute scores for doing a match :-) We can discuss offline to discuss the details. If the control plane sends only one filterChain then there is no need to implement a matcher and we can just error out when multiple are received. Do you agree?

Reply:
okay if we don't need to compare, problem solved. 
I am not sure about it is desired behavior or not. how confident are we that the control plain will only send 1 filterChain (including in near future)?

Reply:
So I am talking to the control plane team to find out how many filterChains I can expect to get. If it is just 1 (or max 2) I don't see why spend so much time on the Comparator. But I have opened this again (marked it not resolved) and will get back to this after addressing all the other comments. By that time I would have some answers and hopefully a clearer idea about which way to go and we can discuss offline if nedeed.

Reply:
So it looks like the planned interface will take some time to be implemented on the control plane side so we still need to be prepared to handle multiple filterChains on our side. But the bottom line is that there will be exactly one filterChain that will be a match for the localAddress where destPort, localIPaddress will match and there is no need to implement either a score or sorting based selection of filterChain. So that is my plan for now: select the filterChain that perfectly matches my localAddress or else report error.

Reply:
So I ended up implementing the `Comparator` as per the suggestion here. PTAL